slack-lint-checks/src/main/java/slack/lint/JsonInflaterMoshiCompatibilityDetector.kt

```diff
@@ -0,0 +1,160 @@
+    return object : UElementHandler() {
+      override fun visitCallExpression(node: UCallExpression) {
+        if (isJsonInflaterInflateOrDeflate(node)) {
+          when (node.methodName) {
+            "inflate" -> validateInflateReturnType(context, node)
+            "deflate" -> validateDeflateArguments(context, node)
+          }
+        }
+      }
+    }
+  }
+
+  private fun isJsonInflaterInflateOrDeflate(node: UCallExpression): Boolean {
+    // Get the method being called
+    val method = node.resolve() ?: return false
+
+    // Check if it's from JsonInflater class
+    val containingClass = method.containingClass?.qualifiedName ?: return false
+    if (containingClass != FQCN_JSON_INFLATER) return false
+
+    // Check if it's an inflate or deflate method
+    return method.name == "inflate" || method.name == "deflate"
+  }
+
+  private fun validateInflateReturnType(context: JavaContext, node: UCallExpression) {
+    // Get the return type of the inflate call
+    val returnType = node.getExpressionType() ?: return
+
+    // Skip checking primitive types and String
+    if (isPrimitiveOrString(returnType)) return
+
+    // Get the class for the return type
+    val returnClass = context.evaluator.getTypeClass(returnType) ?: return
+
+    // Validate if the class is Moshi-compatible
+    validateClassForMoshiCompatibility(context, node, returnClass)
+  }
+
+  private fun isPrimitiveOrString(type: PsiType): Boolean {
+    return type is PsiPrimitiveType || type.canonicalText == "java.lang.String"
+  }
+
+  private fun validateDeflateArguments(context: JavaContext, node: UCallExpression) {
+    val method = node.resolve() ?: return
+    val parameters = method.parameterList.parameters
+    val arguments = node.valueArguments
+
+    for ((index, parameter) in parameters.withIndex()) {
+      if (parameter.name == "value" && index < arguments.size) {
+        val objectArgument = arguments[index]
+        // Get the type of the object being deflated
+        val objectType = objectArgument.getExpressionType() ?: return
+        val objectClass = context.evaluator.getTypeClass(objectType) ?: return
+
+        validateClassForMoshiCompatibility(context, node, objectClass)
+      }
+    }
+  }
+
+  private fun validateClassForMoshiCompatibility(
+    context: JavaContext,
+    node: UCallExpression,
+    classToValidate: PsiClass,
+  ) {
+    if (!isMoshiCompatible(classToValidate)) {
+      context.report(
+        issue = ISSUE_JSON_INFLATER_WITH_MOSHI_INCOMPATIBLE_TYPE,
+        location = context.getLocation(node),
+        message =
+          ISSUE_JSON_INFLATER_WITH_MOSHI_INCOMPATIBLE_TYPE.getBriefDescription(TextFormat.TEXT),
+      )
+    }
+  }
+
+  private fun isMoshiCompatible(psiClass: PsiClass): Boolean {
+    if (!isInstantiable(psiClass)) return false
+
+    return hasPublicNoArgConstructor(psiClass) ||
+      hasJsonClassGenerateAdapter(psiClass) ||
+      hasAdaptedBy(psiClass)
+  }
```

Comment:
I think there is some logic you can reuse in the existing moshi usage checks? Namely there is a check somewhere on JsonClass-annotated class properties to check if their type is a known moshi type

Reply:
are you referring to `hasMoshiAnnotation()`?

Reply:
[3b6262c](https://github.com/slackhq/slack-lints/pull/387/commits/3b6262c1688480d7bf8d1bddd2ac32c6407bf11f)